File: entity.js
// var $jscomp = $jscomp || {};
// $jscomp.scope = {};
// $jscomp.createTemplateTagFirstArg = function (a) {
// 	return a.raw = a
// };
// $jscomp.createTemplateTagFirstArgWithRaw = function (a, b) {
// 	a.raw = b;
// 	return a
// };
var identifier = document.getElementById("entity.identifier"),
	textureFile = document.getElementById("entity.texture_file"),
	modelFile = document.getElementById("entity.model_file"),
	isSpawnable = document.getElementById("entity.is_spawnable"),
	isSummonable = document.getElementById("entity.is_summonable"),
	isExperimental = document.getElementById("entity.is_experimental"),
	spawnEggPanel = document.getElementById("entity.spawn_egg.panel"),
	spawnEggOverlay = document.getElementById("entity.spawn_egg.overlay_color"),
	spawnEggBase = document.getElementById("entity.spawn_egg.base_color"),
	material = document.getElementById("entity.material"),
	baseBeh = document.getElementById("entity.basic_behaviors"),
	genButton = document.getElementById("entity.generate"),
	clearButton = document.getElementById("entity.clear"),
	spawnEggShow = document.getElementById("entity.spawn_egg.show_panel"),
	client = document.getElementsByClassName("client"),
	server = document.getElementsByClassName("server"),
	customBeh = document.getElementById("entity.custom_behaviors"),
	geoID = "",
	select = document.getElementById("pack.type"),
	reader = new FileReader,
	health = document.getElementById("behavior.health"),
	scale = document.getElementById("behavior.scale"),
	collisionH = document.getElementById("behavior.collision_box.height"),
	collisionW = document.getElementById("behavior.collision_box.width"),
	typeFamily = document.getElementById("behavior.type_family"),
	isImmune = document.getElementById("behavior.is_immune");
select.onchange = function () {
	"res" == select.value && "both" != select.value ? ($(".client").show(), $(".server").hide()) : ($(".client").hide(), $(".server").show());
	spawnEggShow.style.display = "none";
	"both" == select.value && ($(".client").show(), $(".server").show(), spawnEggShow.style.display = "block")
};

function main() {
	identifier.value 
		? (
			$("#entity").find("input, select, button").removeAttr("disabled"), 
			spawnEggPanel.style.display = isSpawnable.checked ? "block" : "none"
		) 
		: (
			$("#entity").find("input, select, button").not("[name=identifier]").attr("disabled", "disabled"), 
			spawnEggPanel.style.display = "none"
		);
}
main();

// function getGeo() {
// 	0 < modelFile.files.length && (reader.addEventListener("load", function () {
// 		geoID = JSON.parse(reader.result)["minecraft:geometry"][0].description.identifier
// 	}), reader.readAsText(modelFile.files[0]))
// }
genButton.onclick = function () {
	if (is_valid_identifier(identifier.value)) {
		var uuidobj = new uuid_generator_class,
			jszobject = new JSZip;
		genManifest(uuidobj);
		var manifest = genManifest(uuidobj);
		respack = manifest.rp;
		behpack = manifest.bp;
		var cliententity = {
			format_version: "1.10.0",
			"minecraft:client_entity": {}
		};
		cliententity["minecraft:client_entity"].description = {};
		cliententity["minecraft:client_entity"].description.identifier = identifier.value;
		isSpawnable.checked && (
			cliententity["minecraft:client_entity"].description.spawn_egg = {}, 
			cliententity["minecraft:client_entity"].description.spawn_egg.overlay_color = spawnEggOverlay.value, 
			cliententity["minecraft:client_entity"].description.spawn_egg.base_color = spawnEggBase.value
		);
		var e = textureFile.files;
		for (i = 0; i < e.length; i++) {
			var f = e[i].name.replace(".png", "");
			cliententity["minecraft:client_entity"].description.textures = {};
			cliententity["minecraft:client_entity"].description.textures["default"] = "textures/entity/" + f.toLowerCase();
			"res" !== select.value && "both" !== select.value || jszobject.file(
				"resource_pack/textures/entity/" + (f.toLowerCase() + ".png"), 
				e[i]
			)
		}
		cliententity["minecraft:client_entity"].description.materials = {};
		cliententity["minecraft:client_entity"].description.materials["default"] = material.value;
		0 < modelFile.files.length && (
			cliententity["minecraft:client_entity"].description.geometry = {}, 
			"res" !== select.value && "both" !== select.value || jszobject.file(
				"resource_pack/models/entity/" + modelFile.files[0].name, 
				modelFile.files[0]
			), 
			cliententity["minecraft:client_entity"].description.geometry["default"] = geoID.toLowerCase()
		);
		cliententity["minecraft:client_entity"].description.render_controllers = [];
		cliententity["minecraft:client_entity"].description.render_controllers[0] = "controller.render." + identifier.value.split(":")[1];
		rendercontr = {};
		"beh" !== select.value && (0 < modelFile.files.length && 0 < textureFile.files.length 
			? genRenderControllers(rendercontr) 
			: window.alert('Can\'t generate render controllers, missing either "geometry" or "texture"!'));
		beh = {};
		baseBeh.checked ? genBehBasic(beh) : genBehCustom(beh);
		"res" !== select.value && "both" !== select.value || "beh" === select.value || (
			jszobject.file(
				"resource_pack/manifest.json", 
				JSON.stringify(respack, null, 2)
			), 
			jszobject.file(
				"resource_pack/texts/en_US.lang", 
				"entity." + identifier.value + ".name=" + identifier.value.split(":")[1]
			), 
			jszobject.file(
				"resource_pack/entity/" + identifier.value.split(":")[1] + ".client.json", 
				JSON.stringify(cliententity, null, 2)
			), 
			jszobject.file(
				"resource_pack/render_controllers/" + identifier.value.split(":")[1] + ".render_controllers.json", 
				JSON.stringify(rendercontr, null, 2)
			)
		);
		"beh" !== select.value && "both" !== select.value || "res" === select.value || (
			jszobject.file(
				"behavior_pack/manifest.json", 
				JSON.stringify(behpack, null, 2)
			), 
			jszobject.file(
				"behavior_pack/entities/" + identifier.value.split(":")[1] + ".json", 
				JSON.stringify(beh, null, 2)
			)
		);
		jszobject.generateAsync({
			type: "blob"
		}).then(function (g) {
			"beh" === select.value || "res" === select.value 
				? "res" === select.value 
					? saveAs(g, identifier.value.split(":")[1] + ".rp.mcpack") 
					: saveAs(g, identifier.value.split(":")[1] + ".bp.mcpack") 
				: "both" === select.value && saveAs(g, identifier.value.split(":")[1] + ".mcaddon")
		}, function (g) {
			console.log(g)
		})
	} else window.alert("Invalid Identifier!")
};

function genBehBasic(entityobj) {
	entityobj.format_version = "1.16.0";
	entityobj["minecraft:entity"] = {};
	entityobj["minecraft:entity"].description = {};
	entityobj["minecraft:entity"].description.identifier = identifier.value;
	entityobj["minecraft:entity"].description.is_spawnable = isSpawnable.checked;
	entityobj["minecraft:entity"].description.is_summonable = isSummonable.checked;
	entityobj["minecraft:entity"].description.is_experimental = isExperimental.checked;
	entityobj["minecraft:entity"].components = {};
	entityobj["minecraft:entity"].components["minecraft:health"] = {};
	entityobj["minecraft:entity"].components["minecraft:health"].value = 20;
	entityobj["minecraft:entity"].components["minecraft:physics"] = {}
}

function genBehCustom(entityobj) {
	entityobj.format_version = "1.16.0";
	entityobj["minecraft:entity"] = {};
	entityobj["minecraft:entity"].description = {};
	entityobj["minecraft:entity"].description.identifier = identifier.value;
	entityobj["minecraft:entity"].description.is_spawnable = isSpawnable.checked;
	entityobj["minecraft:entity"].description.is_summonable = isSummonable.checked;
	entityobj["minecraft:entity"].description.is_experimental = isExperimental.checked;
	entityobj["minecraft:entity"].components = {};
	health.value && !isImmune.checked && (
		entityobj["minecraft:entity"].components["minecraft:health"] = {}, 
		entityobj["minecraft:entity"].components["minecraft:health"].value = parseFloat(health.value)
	);
	scale.value && (
		entityobj["minecraft:entity"].components["minecraft:scale"] = {}, 
		entityobj["minecraft:entity"].components["minecraft:scale"].value = parseFloat(scale.value)
	);
	typeFamily.value && (
		entityobj["minecraft:entity"].components["minecraft:type_family"] = {}, 
		entityobj["minecraft:entity"].components["minecraft:type_family"].family = [], 
		entityobj["minecraft:entity"].components["minecraft:type_family"].family = typeFamily.value.split(",")
	);
	isImmune.checked && (
		entityobj["minecraft:entity"].components["minecraft:damage_sensor"] = {}, 
		entityobj["minecraft:entity"].components["minecraft:damage_sensor"].triggers = {}, 
		entityobj["minecraft:entity"].components["minecraft:damage_sensor"].triggers.cause = "all", 
		entityobj["minecraft:entity"].components["minecraft:damage_sensor"].triggers.deals_damage = "false");
	collisionH.value && collisionW.value && (
		entityobj["minecraft:entity"].components["minecraft:collision_box"] = {}, 
		entityobj["minecraft:entity"].components["minecraft:collision_box"].height = parseFloat(collisionH.value), 
		entityobj["minecraft:entity"].components["minecraft:collision_box"].width = parseFloat(collisionW.value)
	)
}

function genRenderControllers(rcobject) {
	rcobject.format_version = "1.10.0";
	rcobject.render_controllers = {};
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]] = {};
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]].geometry = "geometry.default";
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]].materials = [{}];
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]].materials[0]["*"] = "material.default";
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]].textures = [];
	rcobject.render_controllers["controller.render." + identifier.value.split(":")[1]].textures[0] = "texture.default"
}

function genManifest(uuidobj) {
	var behaviorpack = {},
		resourcepack = {};
	behaviorpack.format_version = 2;
	resourcepack.format_version = 2;
	behaviorpack.header = {};
	resourcepack.header = {};
	behaviorpack.header.name = "BP";
	resourcepack.header.name = "RP";
	behaviorpack.header.description = "BP Description";
	resourcepack.header.description = "RP Description";
	behaviorpack.header.uuid = uuidobj.generate_uuid();
	resourcepack.header.uuid = uuidobj.generate_uuid();
	behaviorpack.header.version = [0, 0, 1];
	resourcepack.header.version = [0, 0, 1];
	behaviorpack.header.min_engine_version = [1, 16, 0];
	resourcepack.header.min_engine_version = [1, 16, 0];
	behaviorpack.modules = [{}];
	resourcepack.modules = [{}];
	behaviorpack.modules[0].type = "data";
	resourcepack.modules[0].type = "resources";
	behaviorpack.modules[0].uuid = uuidobj.generate_uuid();
	resourcepack.modules[0].uuid = uuidobj.generate_uuid();
	behaviorpack.modules[0].version = [0, 0, 1];
	resourcepack.modules[0].version = [0, 0, 1];
	"both" === select.value && (
		behaviorpack.dependencies = [{}], 
		resourcepack.dependencies = [{}], 
		behaviorpack.dependencies[0].uuid = resourcepack.header.uuid, 
		resourcepack.dependencies[0].uuid = behaviorpack.header.uuid, 
		behaviorpack.dependencies[0].version = resourcepack.header.version, 
		resourcepack.dependencies[0].version = behaviorpack.header.version
	);
	return {
		rp: resourcepack,
		bp: behaviorpack
	}
}
var is_valid_identifier = verifyIdentifier();

function verifyIdentifier() {
	return function (a, b) {
		b = void 0 === b ? !1 : b;
		return a.match(":") 
			? (a.startsWith("minecraft") || a.startsWith("minecon")) && !1 === b 
				? !1 
				: /[!@#$%^&*()+\-=\[\]{};'"\\|,<>\/?]+/.test(a) 
					? !1 
					: a.startsWith(":") || a.startsWith(".") 
						? !1 
						: !0 
			: !1
	}
};

File: generate_libs/entity.cpp
#include "make_json.h"
#include "uuid_generator.h"
#include "file_management.h"

// ***********************************
// * Note to self: don't get lost :) *
// ***********************************
namespace generatelibs{

class entity{
private:
	bool verifyIdentifier(string identifier){
		regex regex_string("([^!@#$%^&*()+\\-=\\[\\]{};'\"\\\\|,<>\\/?.:][^!@#$%^&*()+\\-=\\[\\]{};'\"\\\\|,<>\\/?]*):([^!@#$%^&*()+\\-=\\[\\]{}:;'\"\\\\|,<>\\/?]+)");
		smatch match_string;
		bool a=regex_match(identifier,match_string,regex_string);
		if (a){
			return (match_string[1].str()!="minecraft"&&match_string[1].str()!="minecon");
		}
		return false;
	}
	string get_ename(string identifier){
		regex regex_string("([^!@#$%^&*()+\\-=\\[\\]{};'\"\\\\|,<>\\/?.:][^!@#$%^&*()+\\-=\\[\\]{};'\"\\\\|,<>\\/?]*):([^!@#$%^&*()+\\-=\\[\\]{}:;'\"\\\\|,<>\\/?]+)");
		smatch match_string;
		bool a=regex_match(identifier,match_string,regex_string);
		if (a){
			return match_string[2];
		}
		return "";
	}
	string strippng(string filename){
		regex regex_string("(.*)\\.png");
		smatch match_string;
		bool a=regex_match(filename,match_string,regex_string);
		if (a){
			return match_string[1].str();
		}
		return "";
	}
	string to_lower(string a){
		int l=a.length();
		string b;
		for (int i=0;i<l;i++){
			if (a[i]>64&&a[i]<91) b+=a[i]+32;
			else b+=a[i];
		}
		return b;
	}
	json_value genManifest(bool select_both){
		uuid_generator uuid;
		json_value bp=json_value(json_object<json_value>{});
		json_value rp=json_value(json_object<json_value>{});
		bp.val_object["format_version"]=2;
		rp.val_object["format_version"]=2;
		bp.val_object["header"]=json_value(json_object<json_value>{});
		rp.val_object["header"]=json_value(json_object<json_value>{});
		bp.val_object["header"].val_object["name"]=string("BP");
		rp.val_object["header"].val_object["name"]=string("RP");
		bp.val_object["header"].val_object["description"]=string("BP Description");
		rp.val_object["header"].val_object["description"]=string("RP Description");
		bp.val_object["header"].val_object["uuid"]=uuid.generate_uuid();
		rp.val_object["header"].val_object["uuid"]=uuid.generate_uuid();
		bp.val_object["header"].val_object["version"]=json_value(vector<json_value>{0,0,1});
		rp.val_object["header"].val_object["version"]=json_value(vector<json_value>{0,0,1});
		bp.val_object["header"].val_object["min_engine_version"]=json_value(vector<json_value>{1,16,0});
		rp.val_object["header"].val_object["min_engine_version"]=json_value(vector<json_value>{1,16,0});
		bp.val_object["modules"]=json_value(vector<json_value>{json_object<json_value>{}});
		rp.val_object["modules"]=json_value(vector<json_value>{json_object<json_value>{}});
		bp.val_object["modules"].val_list[0].val_object["type"]=string("data");
		rp.val_object["modules"].val_list[0].val_object["type"]=string("resources");
		bp.val_object["modules"].val_list[0].val_object["uuid"]=uuid.generate_uuid();
		rp.val_object["modules"].val_list[0].val_object["uuid"]=uuid.generate_uuid();
		bp.val_object["modules"].val_list[0].val_object["version"]=json_value(vector<json_value>{0,0,1});
		rp.val_object["modules"].val_list[0].val_object["version"]=json_value(vector<json_value>{0,0,1});
		if (select_both){
			bp.val_object["dependencies"]=json_value(vector<json_value>{json_object<json_value>{}});
			rp.val_object["dependencies"]=json_value(vector<json_value>{json_object<json_value>{}});
			bp.val_object["dependencies"].val_list[0].val_object["uuid"]=rp.val_object["header"].val_object["uuid"];
			rp.val_object["dependencies"].val_list[0].val_object["uuid"]=bp.val_object["header"].val_object["uuid"];
			bp.val_object["dependencies"].val_list[0].val_object["version"]=rp.val_object["header"].val_object["version"];
			rp.val_object["dependencies"].val_list[0].val_object["version"]=rp.val_object["header"].val_object["version"];
		}
		return json_value(json_object<json_value>{{{"rp",rp},{"bp",bp}}});
	}
	json_value genRenderControllers(string idv1){
		json_value rcobject=json_value(json_object<json_value>{});
		rcobject.val_object["format_version"]=string("1.10.0");
		rcobject.val_object["render_controllers"]=json_value(json_object<json_value>{});
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1]=json_value(json_object<json_value>{});
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1].val_object["geometry"]=string("geometry.default");
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1].val_object["materials"]=json_value(vector<json_value>{json_object<json_value>{}});
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1].val_object["materials"].val_list[0].val_object["*"]=string("material.default");
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1].val_object["textures"]=json_value(vector<json_value>{});
		rcobject.val_object["render_controllers"].val_object[string("controller.render.")+idv1].val_object["textures"].val_list.push_back(string("texture.default"));
		return rcobject;
	}
	json_value genBehBasic(string identifier,bool isSpawnable,bool isSummonable,bool isExperimental){
		json_value entityobj=json_value(json_object<json_value>{});
		entityobj.val_object["minecraft:entity"]=json_value(json_object<json_value>{});
		entityobj.val_object["minecraft:entity"].val_object["description"]=json_value(json_object<json_value>{});
		entityobj.val_object["minecraft:entity"].val_object["description"].val_object["identifier"]=identifier;
		entityobj.val_object["minecraft:entity"].val_object["description"].val_object["is_spawnable"]=isSpawnable;
		entityobj.val_object["minecraft:entity"].val_object["description"].val_object["is_summonable"]=isSummonable;
		entityobj.val_object["minecraft:entity"].val_object["description"].val_object["is_experimental"]=isExperimental;
		entityobj.val_object["minecraft:entity"].val_object["components"]=json_value(json_object<json_value>{});
		entityobj.val_object["minecraft:entity"].val_object["components"].val_object["minecraft:health"]=json_value(json_object<json_value>{});
		entityobj.val_object["minecraft:entity"].val_object["components"].val_object["minecraft:health"].val_object["value"]=20;
		entityobj.val_object["minecraft:entity"].val_object["components"].val_object["minecraft:physics"]=json_value(json_object<json_value>{});
		return entityobj;
	}
public:
	file make(
		string identifier,
		file textureFile,
		file modelFile,
		bool isSpawnable,
		bool isSummonable,
		bool isExperimental,
		string spawnEggOverlay,
		string spawnEggBase,
		string material,
		string geoID,
		string select
	){
		file jszobject=file("",vector<file>{}); // It's not JS but whatever //
		string fn;
		if (verifyIdentifier(identifier)){
			if (select=="beh") jszobject.directory_contents.push_back(file("behavior_pack",vector<file>{}));
			else jszobject.directory_contents.push_back(file("resource_pack",vector<file>{}));
			if (select=="both") jszobject.directory_contents.push_back(file("behavior_pack",vector<file>{}));
			json_value manifest=genManifest(select=="both");
			json_value respack=manifest.val_object["rp"];
			json_value behpack=manifest.val_object["bp"];
			json_value cliententity=json_value(json_object<json_value>{{
				{"format_version",string("1.10.0")},
				{"minecraft:client_entity",json_value(json_object<json_value>{})}
			}});
			cliententity.val_object["minecraft:client_entity"].val_object["description"]=json_value(json_object<json_value>{});
			cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["identifier"]=identifier;
			if (isSpawnable){
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["spawn_egg"]=json_value(json_object<json_value>{});
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["spawn_egg"].val_object["overlay_color"]=spawnEggOverlay;
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["spawn_egg"].val_object["base_color"]=spawnEggBase;
			}
			if (string(textureFile.filename)!=""){
				string f=strippng(textureFile.filename);
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["textures"]=json_value(json_object<json_value>{});
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["textures"].val_object["default"]="textures/entity/"+f;
				if (select!="beh") jszobject.directory_contents[0].directory_contents.push_back(file("textures",vector<file>{file("entity",vector<file>{textureFile})}));
			}
			cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["materials"]=json_value(json_object<json_value>{});
			cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["materials"].val_object["default"]=material;
			if (string(modelFile.filename)!=""){
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["geometry"]=json_value(json_object<json_value>{});
				cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["geometry"].val_object["default"]=to_lower(geoID);
				if (select!="beh") jszobject.directory_contents[0].directory_contents.push_back(file("models",vector<file>{file("entity",vector<file>{modelFile})}));
			}
			cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["render_controllers"]=json_value(vector<json_value>{});
			cliententity.val_object["minecraft:client_entity"].val_object["description"].val_object["render_controllers"].val_list.push_back(string("controller.render.")+get_ename(identifier));
			json_value rendercontr=json_value(json_object<json_value>{});
			if (select!="beh"&&string(modelFile.filename)!=""&&string(textureFile.filename)!=""){
				rendercontr=genRenderControllers(get_ename(identifier));
			}
			// **********************************************
			// * NOTE (for JustAsh): I'm a lazy bum, so     *
			// * I'm not bothering to fix your broken code. *
			// **********************************************
			json_value beh=genBehBasic(identifier,isSpawnable,isSummonable,isExperimental);
			if (select!="beh"){
				jszobject.directory_contents[0].directory_contents.push_back(file("manifest.json",respack.get_repr()));
				jszobject.directory_contents[0].directory_contents.push_back(file("texts",vector<file>{file("en_US.lang",string("entity")+identifier+".name="+get_ename(identifier))}));
				jszobject.directory_contents[0].directory_contents.push_back(file("entity",vector<file>{file(get_ename(identifier)+".client.json",cliententity.get_repr())}));
				jszobject.directory_contents[0].directory_contents.push_back(file("render_controllers",vector<file>{file(get_ename(identifier)+".render_controllers.json",rendercontr.get_repr())}));
			}
			if (select!="res"){
				jszobject.directory_contents[(select=="both")].directory_contents.push_back(file("manifest.json",behpack.get_repr()));
				jszobject.directory_contents[(select=="both")].directory_contents.push_back(file("entities",vector<file>{file(get_ename(identifier)+".json",beh.get_repr())}));
			}
			cout <<jszobject.filename << endl;
			if (select=="beh") fn=(get_ename(identifier)+".bp.mcpack");
			else if (select=="res") fn=(get_ename(identifier)+".rp.mcpack");
			else if (select=="both") fn=(get_ename(identifier)+".mcaddon");
		}
		jszobject.filename=fn;
		return jszobject;
	}
};

}